### PR TITLE
Mypy + Ruff

### PR DESCRIPTION
**Pull request by deedy5**
Mypy check + Ruff lint + Ruff format.

New commands in __`Makefile`__:
1. `make lint`: 
а) Ruff linter checking, 
b) Ruff formatter checking, 
c) Mypy type checking;
2. `make format`: 
a) Ruff linter fix, 
b) Ruff formatter fix.

### CODE REVIEW

**Comment by yifeikong** [comment]
Cool, I have wanted to do this for a long time, but didn’t have the bandwidth.

**Comment by deedy5** [comment]
Mypy swears at the lines in add_reader and add_writer, but on my computer there is no such error. Adding # type: ignore doesn't help for some reason.

Does anyone have this error?
https://github.com/deedy5/curl_cffi/blob/4412a313acc3dd2c594c817f6d4940d3b92207ac/curl_cffi/_asyncio_selector.py#L333

```python3
class AddThreadSelectorEventLoop(asyncio.AbstractEventLoop):
    """Wrap an event loop to add implementations of the ``add_reader`` method family.

    Instances of this class start a second thread to run a selector.
    This thread is completely hidden from the user; all callbacks are
    run on the wrapped event loop's thread.

    This class is used automatically by Tornado; applications should not need
    to refer to it directly.

    It is safe to wrap any event loop with this class, although it only makes sense
    for event loops that do not implement the ``add_reader`` family of methods
    themselves (i.e. ``WindowsProactorEventLoop``)

    Closing the ``AddThreadSelectorEventLoop`` also closes the wrapped event loop.

    """

    # This class is a __getattribute__-based proxy. All attributes other than those
    # in this set are proxied through to the underlying loop.
    MY_ATTRIBUTES = {
        "_real_loop",
        "_selector",
        "add_reader",
        "add_writer",
        "close",
        "remove_reader",
        "remove_writer",
    }

    def __getattribute__(self, name: str) -> Any:
        if name in AddThreadSelectorEventLoop.MY_ATTRIBUTES:
            return super().__getattribute__(name)
        return getattr(self._real_loop, name)

    def __init__(self, real_loop: asyncio.AbstractEventLoop) -> None:
        self._real_loop = real_loop
        self._selector = SelectorThread(real_loop)

    def close(self) -> None:
        self._selector.close()
        self._real_loop.close()

    def add_reader(
        self, fd: _FileDescriptorLike, callback: Callable[..., None], *args: Any  # type: ignore
    ) -> None:  
        return self._selector.add_reader(fd, callback, *args)

    def add_writer(
        self, fd: _FileDescriptorLike, callback: Callable[..., None], *args: Any  # type: ignore
    ) -> None:
        return self._selector.add_writer(fd, callback, *args)

    def remove_reader(self, fd: _FileDescriptorLike) -> bool:
        return self._selector.remove_reader(fd)

    def remove_writer(self, fd: _FileDescriptorLike) -> bool:
        return self._selector.remove_writer(fd)
```

**Comment by deedy5** [comment]
@yifeikong Check it out.

**Comment by deedy5** [comment]
@yifeikong Done

File: pyproject.toml [hunk 1]
```diff
@@ -117,6 +117,17 @@ before-all = "gmake preprocess"
 asyncio_mode = "auto"
 
 
+[tool.ruff]
+line-length = 100
+target-version = "py38"
```
Comment (T-256) [inline pyproject.toml:122]: - I think `target-version` implies from `project.requires-python` by default.

- Why decided to not use default line-length which is 88?

File: tests/threads/test_eventlet.py [hunk 2]
```diff
@@ -2,10 +2,10 @@
 
 eventlet.monkey_patch()
 
-import threading
-import time
+import threading  #  noqa: E402
```
Comment (T-256) [inline tests/threads/test_eventlet.py:5]: use it once at head to skip all E402 violations.

**Review by T-256 — COMMENTED** [review]
Thanks You!

File: .github/workflows/build-and-test.yaml [hunk 3]
```diff
@@ -57,6 +64,11 @@ jobs:
           brew install make automake libtool
           which pipx || brew install pipx && pipx ensurepath
 
+      - name: Lint
```
[the remark this answers is not in the document]
Comment (deedy5) [reply .github/workflows/build-and-test.yaml]: Done

File: pyproject.toml (hunk 1 above)
Comment (deedy5) [reply pyproject.toml]: 1) you are right https://docs.astral.sh/ruff/settings/#target-version. Done.
2) readability - allowing more content on a single line, more complex expressions are not carried over to the following lines.

File: tests/threads/test_eventlet.py (hunk 2 above)
Comment (deedy5) [reply tests/threads/test_eventlet.py]: Done

**Review by T-256 — APPROVED** [review]
LGTM, Thanks!

File: tests/threads/test_eventlet.py [hunk 4]
```diff
@@ -1,12 +1,12 @@
-import eventlet
```
Comment (yifeikong) [inline tests/threads/test_eventlet.py:1]: Same as gevent.

File: tests/threads/test_eventlet.py (hunk 2 above)
Comment (deedy5) [reply tests/threads/test_eventlet.py:5]: Reverted

**Comment by deedy5** [comment]
We can make `strict = true` in project.toml, but we'll have to do a lot more work.

File: tests/threads/test_gevent.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
-from gevent import monkey
```
[the remark this answers is not in the document]
Comment (deedy5) [reply tests/threads/test_gevent.py]: Reversed last commit

File: tests/threads/test_eventlet.py (hunk 4 above)
Comment (deedy5) [reply tests/threads/test_eventlet.py]: Reversed last commit

File: curl_cffi/__init__.py [hunk 6]
```diff
@@ -14,11 +14,9 @@
...
-# This line includes _wrapper.so into the wheel
-from ._wrapper import ffi, lib  # type: ignore
 
-from .const import CurlInfo, CurlMOpt, CurlOpt, CurlECode, CurlHttpVersion, CurlWsFlag
-from .curl import Curl, CurlError, CurlMime
+# This line includes _wrapper.so into the wheel
+from ._wrapper import ffi, lib
 from .aio import AsyncCurl
-
-from .__version__ import __title__, __version__, __description__, __curl_version__
```
Comment (yifeikong) [inline curl_cffi/__init__.py:24]: This line is needed for downstream packages and users.

File: curl_cffi/requests/session.py [hunk 7]
```diff
@@ -369,15 +379,15 @@ def _set_curl_options(
             else:
                 # trick from: https://github.com/yifeikong/curl_cffi/issues/156
                 c.setopt(CurlOpt.LOW_SPEED_LIMIT, 1)
-                c.setopt(CurlOpt.LOW_SPEED_TIME, math.ceil(all_timeout))  # type: ignore
+                c.setopt(CurlOpt.LOW_SPEED_TIME, math.ceil(all_timeout))
 
-        else:
+        elif isinstance(timeout, float):
```
Comment (yifeikong) [inline curl_cffi/requests/session.py:384]:
```suggestion
        elif isinstance(timeout, (int, float)):
```

File: curl_cffi/__init__.py (hunk 6 above)
Comment (deedy5) [reply curl_cffi/__init__.py:24]: Done

File: curl_cffi/requests/session.py (hunk 7 above)
Comment (deedy5) [reply curl_cffi/requests/session.py]: Done

File: curl_cffi/requests/headers.py [hunk 8]
```diff
@@ -11,8 +11,6 @@
     typing.Mapping[bytes, bytes],
     typing.Sequence[typing.Tuple[str, str]],
     typing.Sequence[typing.Tuple[bytes, bytes]],
-    typing.Sequence[str],
```
Comment (yifeikong) [inline curl_cffi/requests/headers.py:14]: `list[str|bytes]` is actually used in the `__init__` function. See line 90

File: curl_cffi/aio.py [hunk 9]
```diff
@@ -251,3 +210,45 @@ def set_exception(self, curl: Curl, exception):
     def setopt(self, option, value):
         """Wrapper around curl_multi_setopt."""
         return lib.curl_multi_setopt(self._curlm, option, value)
+
+
+@ffi.def_extern()
+def timer_function(curlm, timeout_ms: int, clientp: AsyncCurl):
```
Comment (yifeikong) [inline curl_cffi/aio.py:216]: I don't quite get why these functions are moved in the end. Actually, we can annotate it as `"AsyncCurl"` even if it's put in the beginning.

File: curl_cffi/requests/headers.py (hunk 8 above)
Comment (deedy5) [reply curl_cffi/requests/headers.py:14]: Done

File: curl_cffi/aio.py (hunk 9 above)
Comment (deedy5) [reply curl_cffi/aio.py]: Done

**Comment by deedy5** [comment]
@yifeikong :checkered_flag:

**Pull request merged**